### ISSUE TITLE
Service Providers boot() method

### ISSUE THREAD

**Issue by Ifnot** [adonisjs/fold#2]
Hi,

Can we make a boot() method on service providers ?

It can be really usefull. We already have a the `register()` method where we can register biding. Next, the `boot()` method we can trigger actions from other items in the IoC.

It will increase the ability to extend the functionality of the framework using modules.

The `boot()` method should be triggered by the loader just after all `register()` are called and IoC full of binds.

**Comment by thetutlage** [comment]
Can u come up with some practical examples of where it can be useful?

**Comment by Ifnot** [comment]
Thanks for your fast response.

I am making a Electron wrapper for adonis. It will bring some magic stuff for making cross plateform apps with less code. I am using the Laravel approach (as i am initially a PHP dev) witch is :
- Make a new package with a service provider
- Bind your public services using the IoC in the `register()` method
- Call you logic (listening for event, gathering some config datas). In my case : listen for App events (the `App` IoC should be set) and spawn an electron app.

This is the main function for "do something when the app and biding is ok" from unknown source like community modules.

Some good refs :
- https://laravel.com/docs/master/providers#the-boot-method
- https://laracasts.com/discuss/channels/general-discussion/difference-between-boot-and-register-method

**Comment by thetutlage** [comment]
Yeah I understand your point of view, so it will be a method free from the lifecycle of the application. Which means

1. You cannot make asynchronous calls from the boot method.
2. Application itself will not wait for the `boot` methods to finish. It will start doing it's own job and call the `boot` method parallelly.

It does solves the problem of binding to events etc. But I am not sure if there is any requirement to do things, which are required for application to run within the `boot` method. For example-

You are trying to call some api service and get configuration which is required for application to run. As application is not dependent upon the `boot` method, it will start running without the configuration you need and it will break.

**Comment by Ifnot** [comment]
I dont think it should be free from the lifecycle,

Where PHP is full sync, the `boot()` can do actions required by the main application. In javascript I think the `boot()` should be sync for allowing more complex user cases.

With Laravel i created tens of modules doing some interesting logic on `boot()` affecting the rest of the app. Finally, In sync it is our job to make it fastest as possible (otherwise make some async for not slowing the start of the app).
